packages/fsapp/src/beta-app/router/context/button-context.tsx

```diff
@@ -1,42 +1,65 @@
+  onPress(buttonId: string, componentId: string, callback: () => void): () => void;
 }
 
 export const ButtonContext = createContext<IButtonContext>({
-  onPress: () => undefined
+  onPress: () => () => undefined
 });
 export const useButtons = () => useContext(ButtonContext);
 
 export const ButtonProvider: FC = ({ children }) => {
-  const listeners = useMemo(() => new Map<string, (() => void)[]>(), []);
+  const listenerRepo = useMemo(() => new Map<string, Map<string, () => void>>(), []);
+
   const onPress = useCallback(
-    ((buttonId: string, componentIDOrCallback: string | (() => void), callback: () => void) => {
-      const componentID =
-        typeof componentIDOrCallback === 'string' ? componentIDOrCallback : undefined;
+    ((buttonId: string, componentIdOrCallback: string | (() => void), callback: () => void) => {
+      const componentId =
+        typeof componentIdOrCallback === 'string' ? componentIdOrCallback : undefined;
       const actualCallback =
-        typeof componentIDOrCallback === 'function' ? componentIDOrCallback : callback;
+        typeof componentIdOrCallback === 'function' ? componentIdOrCallback : callback;
+
+      const listenersId = [buttonId, ...(componentId ? [componentId] : [])].join('-');
+      const listenerId = uniqueId(listenersId);
+      const listeners =
+        listenerRepo.get(listenersId) ?? listenerRepo.set(listenersId, new Map()).get(listenersId);
+      listeners?.set(listenerId, actualCallback);
 
-      const id = [buttonId, ...(componentID ? [componentID] : [])].join('-');
-      const existingListeners = listeners.get(id);
-      listeners.set(id, [...(existingListeners ?? []), actualCallback]);
+      return () => {
+        listeners?.delete(listenerId);
+      };
     }) as IButtonContext['onPress'],
     []
   );
 
   useEffect(() => {
-    Navigation.events().registerNavigationButtonPressedListener(({ buttonId, componentId }) => {
-      listeners.get(buttonId)?.forEach(callback => callback());
-      listeners.get(`${buttonId}-${componentId}`)?.forEach(callback => callback());
-    });
+    const subscription = Navigation.events().registerNavigationButtonPressedListener(
+      ({ buttonId, componentId }) => {
+        listenerRepo.get(buttonId)?.forEach(callback => callback());
+        listenerRepo.get(`${buttonId}-${componentId}`)?.forEach(callback => callback());
+      }
+    );
+
+    return () => subscription.remove();
   }, []);
 
-  return (
-    <ButtonContext.Provider value={{ onPress }}>
-      {children}
-    </ButtonContext.Provider>
-  );
+  return <ButtonContext.Provider value={{ onPress }}>{children}</ButtonContext.Provider>;
+};
+
+export const useButtonEffect = (
+  callback: () => void,
+  ids: [buttonId: string] | [buttonId: string, componentId: string]
+) => {
+  const buttons = useButtons();
+
+  useEffect(() => {
+    if (ids.length === 1) {
+      return buttons.onPress(...ids, callback);
+    } else {
+      return buttons.onPress(...ids, callback);
```

Comment:
just out of curiosity, is there a reason for having the exact same code in both the `if` and `else`?

Reply:
This is for TypeScript. Because `onPress()` has an overload TypeScript needs to know which type to check against at design time thus narrowing down the type and checking each individually is required to keep TypeScript happy.

Maybe in a future version of TypeScript it will be able to check multiple overloads at once.